### ISSUE TITLE
[explicit-function-return-type] option to allow functions that return as const

### ISSUE THREAD

**Issue by dev054** [typescript-eslint/typescript-eslint#2964]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": "error"
  }
}
```

```TS
function setup() {
  // variable declarations and real names omitted for brevity.
  return { a, b, c, d } as const;
}
```

**Expected Result**
I want to propose a new option to avoid specify type when the function's return is `const`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | 4.1.0 |
| `@typescript-eslint/parser`        | 4.1.0 |
| `TypeScript`                       | 3.9.7 |
| `ESLint`                           | 7.12.1 |

**Comment by bradzacher** [comment]
Thank you for the suggestion!

The rule doesn't check any return statements on purpose.

Checking return statements makes things very complicated very quickly as you have to factor in all of the branches amongst other things.
Only supporting some cases makes the rule a lot more unpredictable and harder to understand - which causes people to lose trust in it.

Additionally there are no options which can ignore function declarations - only expressions.

For these reason - we won't be adding an option for this.
